iron-list.html

```diff
@@ -989,6 +1058,139 @@
           this.scrollToIndex(this.firstVisibleIndex);
         }
       });
+    },
+
+    _getModelFromItem: function(item) {
+      var vidx = this._collection.getKey(item);

```

Comment:
These changes are mistaking the collection key for the array index (virtual index), which is wrong.  The initial value for an array's keys _will_ be the index, which is likely why your testing works, but as items are added/removed, the key will _not_ follow the index.  You should think of "key" as a unique ID that increments for each new item added to the array, regardless of where it is added.  If you unshift an item onto a 3-item array, the keys will now be [3, 0, 1, 2].

As such, there is no O(1) operation for going from item --> array index; that requires an `indexOf`.  If you re-implement this using `indexOf`, you should take care where it's called.

An alternative is to change state in the array-selector (which maintains O(1) operations for select/deselect), and then "refresh" selected state by iterating all physicalItems and updating selected state (since that's just an O(n-of-selectedItems + n-of-physicalItems), which should be much better than O(n-of-selectedItems * n-of-virtualItems)).